src/cloud-sql-instance.ts

```diff
@@ -98,60 +107,153 @@ export class CloudSQLInstance {
+  // used to create new connections to a Cloud SQL instance. It throws in
+  // case any of the internal steps fails.
+  private async performRefresh(): Promise<RefreshResult> {
     const rsaKeys: RSAKeys = await generateKeys();
     const metadata: InstanceMetadata =
       await this.sqlAdminFetcher.getInstanceMetadata(this.instanceInfo);
 
-    this.ephemeralCert = await this.sqlAdminFetcher.getEphemeralCertificate(
+    const ephemeralCert = await this.sqlAdminFetcher.getEphemeralCertificate(
       this.instanceInfo,
       rsaKeys.publicKey,
       this.authType
     );
-    this.host = selectIpAddress(metadata.ipAddresses, this.ipType);
-    this.privateKey = rsaKeys.privateKey;
-    this.serverCaCert = metadata.serverCaCert;
+    const host = selectIpAddress(metadata.ipAddresses, this.ipType);
+    const privateKey = rsaKeys.privateKey;
+    const serverCaCert = metadata.serverCaCert;
+
+    const currentValues = {
+      ephemeralCert: this.ephemeralCert,
+      host: this.host,
+      privateKey: this.privateKey,
+      serverCaCert: this.serverCaCert,
+    };
+
+    const nextValues = {
+      ephemeralCert,
+      host,
+      privateKey,
+      serverCaCert,
+    };
+
+    // In the rather odd case that the current ephemeral certificate is still
+    // valid while we get an invalid result from the API calls, then preserve
+    // the current metadata.
+    if (this.isValid(currentValues) && !this.isValid(nextValues)) {
+      return currentValues as RefreshResult;
+    }
 
-    this.scheduledRefreshID = setTimeout(() => {
-      this.refresh();
-    }, getRefreshInterval(this.ephemeralCert.expirationTime));
+    return nextValues;
+  }
+
+  private isValid({
+    ephemeralCert,
+    host,
+    privateKey,
+    serverCaCert,
+  }: Partial<RefreshResult>): boolean {
+    if (!ephemeralCert || !host || !privateKey || !serverCaCert) {
+      return false;
+    }
+    return isExpirationTimeValid(ephemeralCert.expirationTime);
+  }
+
+  private updateValues(nextValues: RefreshResult): void {
+    const {ephemeralCert, host, privateKey, serverCaCert} = nextValues;
+
+    this.ephemeralCert = ephemeralCert;
+    this.host = host;
+    this.privateKey = privateKey;
+    this.serverCaCert = serverCaCert;
+  }
+
+  private scheduleRefresh(): void {
+    const refreshInterval = getRefreshInterval(
+      /* c8 ignore next */
+      String(this.ephemeralCert?.expirationTime)
+    );
+
+    this.scheduledRefreshID = setTimeout(() => this.refresh(), refreshInterval);
   }
 
   cancelRefresh(): void {
     if (this.scheduledRefreshID) {
       clearTimeout(this.scheduledRefreshID);
     }
+    this.scheduledRefreshID = null;
+  }
+
+  // Mark this instance as having an active connection. This is important to
+  // ensure any possible errors thrown during a future refresh cycle should
+  // not be thrown to the final user.
+  setStablishedConnection(): void {
+    this.stablishedConnection = true;
```

Comment:
nit: typo, should be established
```suggestion
  setEstablishedConnection(): void {
    this.establishedConnection = true;
```

Reply:
good catch! there are a few more occurences, will commit it! thanks!